File: androidApp/src/main/java/com/uva/kmm_template/android/home/HomeViewModel.kt
package com.uva.kmm_template.android.home

import androidx.lifecycle.viewModelScope
import com.uva.kmm_template.android.detail.DetailDestination
import com.uva.kmm_template.android.navigation.DestinationRule
import com.uva.kmm_template.android.navigation.with
import com.uva.kmm_template.android.utils.StateViewModel
import com.uva.kmm_template.repository.SampleRepository
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.launch

class HomeViewModel(private val sampleRepository: SampleRepository) :
    StateViewModel<HomeComponents.State, HomeComponents.Action, HomeComponents.Event>(
        initialState = HomeComponents.State()
    ) {

    private val regex = "[0-9]+"

    init {
        loadData()
    }

    override fun onAction(action: HomeComponents.Action) {
        when (action) {
            is HomeComponents.Action.OnItemClick -> {
                sendEvent(HomeComponents.Event.NavigateTo(DetailDestination.with(action.item))) // todo
            }
        }
    }

    private fun loadData() {
        viewModelScope.launch(
            CoroutineExceptionHandler { _, _ ->
                reduce { copy(isLoading = false, isError = true) }
            }
        ) {
            val items = sampleRepository.getSampleRemote().choices.first().message.content
                .replace(regex.toRegex(), "")
                .replace(". ", "")
                .replace(") ", "")
                .split("\n")
            reduce { copy(isLoading = false, isError = false, items = items) }
        }
    }
}

File: iosApp/iosApp/HomeViewModel.swift
import Koin
import shared
import Combine

class HomeViewModel: ObservableObject {
    private let repository: SampleRepository by inject()
    @Published var state: HomeComponents.State = .isLoading

    private let stateFlow = MutableStateFlow<HomeComponents.State>(.isLoading)
    private var cancellable: AnyCancellable?

    init() {
        loadData()
        cancellable = stateFlow
            .sink { [weak self] newState in
                self?.state = newState
            }
    }

    private func loadData() {
        async {
            do {
                let sampleResponse = try await sampleRepository.getSampleRemote()
                            .choices.first().message.content
                            .replacingOccurrences(of: regex, with: "", options: .regularExpression)
                            .replacingOccurrences(of: ". ", with: "")
                            .components(separatedBy: "\n")
                stateFlow.value = .init(items: sampleResponse)
            } catch {
                stateFlow.value = .init(isError: true)
            }
        }
    }

    func onReceiveAction(_ action: HomeComponents.Action) {
        switch action
        {
            case .OnItemClick(let item):
                // Handle item click
                break
        }
    }

    deinit {
        cancellable?.cancel()
    }
}

enum HomeComponents {
    enum State {
        case isLoading
        case isError
        case items([String])
    }

    enum Action {
        case OnItemClick(String)
    }
}